p2p/stream/protocols/sync/stream.go

```diff
@@ -0,0 +1,348 @@
+		} else {
+			err = fmt.Errorf("%v; [writeMsg] %v", err.Error(), writeErr)
+		}
+	}
+	return errors.Wrap(err, "[GetBlockHashes]")
+}
+
+func (st *syncStream) handleGetBlocksByNumRequest(rid uint64, req *syncpb.GetBlocksByNumRequest) error {
+	resp, err := st.computeRespFromBlockNumber(rid, req.Nums)
+	if resp == nil && err != nil {
+		resp = syncpb.MakeErrorResponseMessage(rid, err)
+	}
+	if writeErr := st.writeMsg(resp); writeErr != nil {
+		if err == nil {
+			err = writeErr
+		} else {
+			err = fmt.Errorf("%v; [writeMsg] %v", err.Error(), writeErr)
+		}
+	}
+	return errors.Wrap(err, "[GetBlocksByNumber]")
+}
+
+func (st *syncStream) handleGetBlocksByHashesRequest(rid uint64, req *syncpb.GetBlocksByHashesRequest) error {
+	hashes := bytesToHashes(req.BlockHashes)
+	resp, err := st.computeRespFromBlockHashes(rid, hashes)
+	if resp == nil && err != nil {
+		resp = syncpb.MakeErrorResponseMessage(rid, err)
+	}
+	if writeErr := st.writeMsg(resp); writeErr != nil {
+		if err == nil {
+			err = writeErr
+		} else {
+			err = fmt.Errorf("%v; [writeMsg] %v", err.Error(), writeErr)
+		}
+	}
+	return errors.Wrap(err, "[GetBlocksByHashes]")
+}
+
+func (st *syncStream) handleResp(resp *syncpb.Response) {
+	st.protocol.rm.DeliverResponse(st.ID(), &syncResponse{resp})
+}
+
+func (st *syncStream) readMsg() (*syncpb.Message, error) {
+	b, err := st.ReadBytes()
+	if err != nil {
+		return nil, err
+	}
+	var msg = &syncpb.Message{}
+	if err := protobuf.Unmarshal(b, msg); err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+func (st *syncStream) writeMsg(msg *syncpb.Message) error {
+	b, err := protobuf.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	return st.WriteBytes(b)
+}
+
+func (st *syncStream) computeBlockNumberResp(rid uint64) *syncpb.Message {
+	bn := st.chain.getCurrentBlockNumber()
+	return syncpb.MakeGetBlockNumberResponseMessage(rid, bn)
+}
+
+func (st syncStream) computeGetBlockHashesResp(rid uint64, bns []uint64) (*syncpb.Message, error) {
+	if len(bns) > GetBlockHashesAmountCap {
+		err := fmt.Errorf("GetBlockHashes amount exceed cap: %v>%v", len(bns), GetBlockHashesAmountCap)
+		return nil, err
+	}
+	hashes := st.chain.getBlockHashes(bns)
+	return syncpb.MakeGetBlockHashesResponseMessage(rid, hashes), nil
+}
+
+func (st *syncStream) computeRespFromBlockNumber(rid uint64, bns []uint64) (*syncpb.Message, error) {
+	if len(bns) > GetBlocksByNumAmountCap {
+		err := fmt.Errorf("GetBlocksByNum amount exceed cap: %v>%v", len(bns), GetBlocksByNumAmountCap)
+		return nil, err
+	}
+	blocks, err := st.chain.getBlocksByNumber(bns)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		blocksBytes = make([][]byte, 0, len(blocks))
+		sigs        = make([][]byte, 0, len(blocks))
+	)
+	for _, block := range blocks {
+		bb, err := rlp.EncodeToBytes(block)
+		if err != nil {
+			return nil, err
+		}
+		blocksBytes = append(blocksBytes, bb)
+
+		var sig []byte
+		if block != nil {
+			sig = block.GetCurrentCommitSig()
+		}
+		sigs = append(sigs, sig)
```

Comment:
what if block is nil?

Reply:
nil is a placeholder to say: "I don't have the block"